### ISSUE TITLE
LLM Fine-Tuning Dataset Generation

### ISSUE THREAD

**Issue by Ishwarpatra** [Ishwarpatra/toc_aiagent#5]
###  ENGINEERING DIRECTIVE: LLM Fine-Tuning Dataset Generation

**Objective**
Create an automated script (generate_finetune_dataset.py) that extracts perfectly verified DFA generations from our batch_verify.py QA pipeline and formats them into a .jsonl dataset. This dataset will be used to fine-tune qwen2.5-coder:1.5b via LoRA, effectively eliminating its tendency to hallucinate JSON schemas and logic structures.

 **Insight: How to Approach This**
We do not need to manually label data. Our pipeline already contains an OracleSolver that mathematically guarantees if a DFA is 100% correct. We will use the pipeline as a Synthetic Data Generator.

**You need to extract data to teach the model three distinct skills:**

- [ ] The Analyst (NL -> LogicSpec): Translating raw text into our intermediate JSON schema.
- [ ] The Architect (LogicSpec -> DFA): Translating atomic rules into valid DFA state machines with trap states (q_dead).
- [ ] The Repair Engine (Broken DFA + Error -> Fixed DFA): Teaching the model how to fix its own mistakes when the Validator catches a structural error.

 The Best Way to Do It (Architecture Guidelines)
1. Script Location & Integration
Create the script at backend/scripts/generate_finetune_dataset.py. Do not alter the core engine to do this; read from the artifacts generated by run_qa_pipeline.py (specifically qa_results.csv and the persistent diskcache).

2. Data Formatting Standard (ChatML)
Qwen2.5 natively expects the ChatML format. Structure your output .jsonl file strictly like this:

> JSON
> 
> {
>   "messages": [
>     {"role": "system", "content": "You are a Theory of Computation AI. Output strict, valid JSON."},
>     {"role": "user", "content": "Design an atomic DFA for alphabet ['0', '1'] that exactly matches '101'."},
>     {"role": "assistant", "content": "{\"states\": [\"q0\", \"q1\", \"q2\", \"q3\", \"q_dead\"], \"alphabet\": [\"0\", \"1\"], ... }"}
>   ]
> }

3. Execution Steps

Load qa_results.csv using pandas or Python's built-in csv.

Filter for rows where status == "PASS".

Fetch the exact prompt (User) and the final accepted dfa_json (Assistant).

Write out to backend/output/qwen_toc_finetune.jsonl.

Bonus points: Introduce a --split flag to automatically generate an 80/20 train.jsonl and eval.jsonl.

 **Resources & Documentation**
Internal Codebase References
backend/scripts/batch_verify.py: Look at how _emit_summary() parses results. You can hook into the exact same self.results array.

backend/python_imply/core/models.py: Reference the DFA and LogicSpec Pydantic models to understand the exact JSON structure the model must learn to output.

External Resources for Fine-Tuning
Unsloth (Highly Recommended): The fastest and easiest way to fine-tune Qwen2.5-Coder-1.5b locally.

Doc: [Unsloth GitHub / Qwen Examples](https://github.com/unslothai/unsloth)

Why: It uses 70% less VRAM and handles ChatML formatting natively.

**Qwen2.5 Official Docs:**

Doc: [Qwen2.5 Prompt Formatting](https://qwen.readthedocs.io/en/latest/)

Hugging Face TRL (Supervised Fine-Tuning):

Doc: [SFTTrainer Documentation](https://huggingface.co/docs/trl/main/en/sft_trainer)

**Definition of Done**
- [ ] generate_finetune_dataset.py is merged into main.

- [ ] Running python generate_finetune_dataset.py --input qa_results.csv outputs a valid train.jsonl containing at least 1,000 ChatML-formatted examples.

- [ ] The generated JSON strings pass a basic json.loads() validation check before being written to the file to ensure no malformed data enters the training set.
